question-servers/freeformPythonLib/prairielearn.py

```diff
@@ -762,12 +762,8 @@ def string_to_2darray(s, allow_complex=True):
         #   Example:
         #       is '[1 - 2j]' the same as '[1 -2j]' or '[1-2j]'
 
-        # Return error if there are any commas
-        if ',' in s:
-            return (None, {'format_error': 'Commas cannot be used as delimiters in an expression with single brackets.'})
-
-        # Split on semicolon
-        s = s.split(';')
+        # Split on semicolon (treat comma as space)
+        s = s.replace(',', ' ').split(';')
```

Comment:
Would it be more robust to do check whether the entire original string contains a comma and, if so, split all rows on comma and otherwise split on space? That would reject `[1,,,,, 2, 3; 4, 5, 6]`.

Reply:
Without adding more complicated logic, that would also reject stuff like `[1, 2, 3; 4 5 6]` (mixing spaces and commas which MATLAB allows). Which limitation do you prefer?

Reply:
Oh! I didn't realize Matlab would accept that. How horrible! 😄 

In that case, can't you first split on semi-colon, then for each row check whether it contains a comma and split on comma or space? I agree this is slightly more complex than your current code, but you could write it like:
```
n = len(s[0].split(',' if ',' in s[0] else None))

s_row = s[i].split(,' if ',' in s[0] else None)
```

Reply:
So my original counter-example didn't make this clear (edited later), but MATLAB allows mixing commas/spaces even within rows. I can try to make PL's parser mimic MATLAB, but I wasn't sure if that's where we wanted to go with it. Sounds like that's what you prefer?

Reply:
Does python allow splits with regex?  Could we just split each row whenever it literally sees a comma or a space(s)?

Reply:
Yep if we go the true-compat route, I was planning to use `re`.

Reply:
I feel like I've bikeshedded this PR to death, but I guess I'd agree that we should just `re` it. Does `re.split(r'\s*(\s|,)\s*)` do what we want?

Reply:
I think this will break if the student has newlines in a Matlab matrix definition as `\s` matches those. So it's important to split by `;` or `\n` first. I'd need to read the Matlab specification to see if there are other edgecases. @tdy Does this cover it?

Reply:
@echuber which "this" are you referencing? The old commit, matt's regex, or the newest commit?

Reply:
Sorry, Matt's regex in the context of this thread.

Reply:
The latest revision splits into cols after first splitting into rows.

Reply:
I can't tell from the code whether it correctly handles the case where students had literally put a newline in their matrix entry. Does the input element prevent that? Octave allows this:

```
>> m = [1 2, 3; 4, 5 6
7 8 9]
m =

   1   2   3
   4   5   6
   7   8   9
```

Reply:
I don't know about the bigger picture, but it looks like `string_to_2darray` was written solely to parse `pl-matrix-input`, which uses `<input>` and makes newlines moot.

Reply:
If there's a potential this could be used some context in the future where that would be a problem, maybe we should also do splitting by `\n` here then, and check that the number of columns per row is consistent (if not already doing so)

Reply:
sounds good